h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,701 @@
+    var referencesLength = (model.references || []).length;
+    var tagsLength = (model.tags || []).length;
+    return (targetLength && !referencesLength && !(model.text || tagsLength));
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#isPrivate
+    * @returns {boolean} True if the annotation is private to the current user.
+    */
+  vm.isPrivate = function() {
+    return permissions.isPrivate(vm.annotation.permissions, model.user);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#isShared
+    * @returns {boolean} True if the annotation is shared (either with the
+    * current group or with everyone).
+    */
+  vm.isShared = function() {
+    return permissions.isShared(vm.annotation.permissions, model.group);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#setPrivacy
+    *
+    * Set the privacy settings on the annotation to a predefined
+    * level. The supported levels are 'private' which makes the annotation
+    * visible only to its creator and 'shared' which makes the annotation
+    * visible to everyone in the group.
+    *
+    * The changes take effect when the annotation is saved
+    */
+  vm.setPrivacy = function(privacy) {
+    // When the user changes the privacy level of an annotation they're
+    // creating or editing, we cache that and use the same privacy level the
+    // next time they create an annotation.
+    // But _don't_ cache it when they change the privacy level of a reply.
+    if (!model.references) {  // If the annotation is not a reply.
+      permissions.setDefault(privacy);
+    }
+    if (privacy === 'private') {
+      vm.annotation.permissions = permissions.private();
+    } else if (privacy === 'shared') {
+      vm.annotation.permissions = permissions.shared(model.group);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotaitonController#hasContent
+    * @returns {boolean} True if the currently edited annotation has
+    *          content (ie. is not just a highlight)
+    */
+  vm.hasContent = function() {
+    var textLength = (vm.annotation.text || '').length;
+    var tagsLength = (vm.annotation.tags || []).length;
+    return (textLength > 0 || tagsLength > 0);
+  };
+
+  /**
+    * @returns {boolean} True if this annotation has quotes
+    */
+  vm.hasQuotes = function() {
+    return vm.annotation.target.some(function(target) {
+      return target.selector && target.selector.some(function(selector) {
+        return selector.type === 'TextQuoteSelector';
+      });
+    });
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#authorize
+    * @param {string} action The action to authorize.
+    * @returns {boolean} True if the action is authorized for the current user.
+    * @description Checks whether the current user can perform an action on
+    * the annotation.
+    */
+  vm.authorize = function(action) {
+    if (model === null) {
```

Comment:
Reading the code above, it doesn't look possible for `model` to be null here, that check should be removed.